### DISCUSSION TITLE
ESPNow Send Issue - requires WLAN.connect() ???

### DISCUSSION THREAD

**Discussion by case485** [micropython/micropython#11814] [General]
I have a ESP32-Wroom-32 Devkitv1 set of boards, I can establish and send/receive ESPNow messages but ONLY when I add the WLAN.connect(“ssid”) function to the flow. After the WLAN.connect(“ssid”) I still call the WLAN.disconnect() function but with out the connect statement I can not send/receive. 

import network
import espnow
import time

sta = network.WLAN(network.STA_IF) 
sta.active(True)
sta.connect(“MySSID”) <——— Only works if I add this connect function
sta.disconnect() 
e = espnow.ESPNow() 
e.active(True)
peer = b'\xe8\x68\xe7\x4e\xbb\x19' 
e.add_peer(peer) 

print("Starting...") 
e.send(peer1, "Test Message", True)

**Comment by davefes** [comment]
The WiFi needs to be active for the ESPNow interface to send and receive.  Calling disconnect when you are connected only makes sure that you are really disconnected.
If you dig into Glenn's wifi.py you will see that a disconnect() is used to ensure the WiFi is in a known state, at start-up
https://github.com/glenn20/micropython-espnow-utils/blob/main/src/wifi.py

**Reply by case485** [reply]
Thanks for the reply, I read through and understand the need for the disconnect() call but I don't understand why I need to call a wlan.connect("ssid") before the disconnect to make ESPNow work. I wanted to host this sensor remotely with no access to a Wi-Fi signal so it will ultimately hang on the connect() function.

**Reply by davefes** [reply]
Fortunately I said "active" and not "connected".

My remotes are configured like this:
```py
    w0 = network.WLAN(network.STA_IF)
    print (w0.config('channel'))
    w0.active(True)
    print (w0.config('mac'))
    w0.config(channel=6)
    print (w0.config('channel'))

    e0 = espnow.ESPNow()
    e0.active(True)

    e0.add_peer(repeater_mac)
    print (e0.get_peers())
```
Maybe, something similar for the "local"end works, all my ESPNow links talk to routers after gathering all the information.  You probably have to do something extra at the local-end.

**Reply by davefes** [reply]
If you go here:
https://github.com/davefes/ESPNow-Range-Testing
there is a stand-alone sender/receiver.

There have been a lot of changes in the last year, so get the latest docs.

**Comment by glenn20** [comment]
That is odd. I have never found it necessary to call `sta.connect()/sta.disconnect()` to use ESPNow.

Please note that if you only soft reset the ESP32, the wifi device is not reset and your device will remain configured as it was before the soft reset. This can lead to some surprising behaviour, including remining connected to an AP, or running on an unexpected channel.

I suggest to try power cycling the device (or calling `machine.reset()`) to hard reset the device then see if your code works without the `sta.connect()/sta.disconnect()` sequence.

**Reply by glenn20** [reply]
Oh, and there is another possibility as well (hard to tell without knowing more about your setup):
- After a hard reset, the radio on the esp32 is set to wifi channel 1. It **may** be the case that the peer devices you are testing with may be running on a different channel, ie. the same channel as your wifi router, and when you connect/disconnect() from the AP, you are changing your device to the same channel as your AP.

Because of the common traps with esp32/8266 devices being in an unexpected state after starting micropython (after soft reset), I have written a utility to properly initialise the wifi radio settings (see the `wifi.py` module at https://github.com/glenn20/micropython-espnow-utils). You can use that or take a look to see what you should do to guarantee the devices is correctly initialised every time micropython starts.

...and the `wifi.status()` can be very useful for printing out all the relevant wifi settings to check everyhting is as you expect (especially the channel numbers for all your devices).

**Comment by case485** [comment]
SOLVED: Looks like it was the channel issue when I added, w0.config(channel=6), the transmission works great. Thanks for all the help. 

import network
import espnow
import time


peer = b'\xa0\xb7eXv8'
w0 = network.WLAN(network.STA_IF)
print (f"Channel: {w0.config('channel')}")
w0.active(True)
print (f"Mac: {w0.config('mac')}")
w0.config(channel=6)  <-------------------------------Works when added
print (w0.config('channel'))
e0 = espnow.ESPNow()
e0.active(True)
e0.add_peer(peer)
print("Starting...")   
e0.send(peer, "Testing", True)

**Reply by glenn20** [reply]
Excellent. This suggest your other espnow devices are running on channel 6 (probably because they are also (or have been) connected to your wifi router - which is running on channel 6). If you hard reset all your devices and don't connect any of them to wifi and **don't set the channel** they will all run on channel 1 by default, but if any of them have previously been connected to your wifi and then soft reset, they will remain on channel 6.

UPDATE: The above statement is true for ESP32 devices. ESP8266 devices remember their last configuration even after hard resets.
